### PR TITLE
Make matrix script process any matrix items

### PR DESCRIPTION
**Pull request by nineinchnick**
## Description

When #10323 introduced a script to filter out non-impacted modules from the the CI test matrix, this script was intentionally very simple and only handled the currently existing matrix. This PR extends it to:
* allow processing any keys in the `include` section, not just currently used ones
* process any matrix items, instead of only processing the `include` section

This allows future contributors to adjust the CI matrix as they see fit, without having to figure out how it's different than the regular GHA matrix.

> Is this change a fix, improvement, new feature, refactoring, or other?
CI tools improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
CI scripts

> How would you describe this change to a non-technical end user or system administrator?
n/a

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Review by MiguelWeezardo — APPROVED** [review]

**Review by hashhar — COMMENTED** [review]
Question: Is it necessary to expand the matrix? Can we just passthrough the matrix as it is and let GHA handle the expansion?

e.g. in current impl we don't correctly handle include where it matches an already existing entry in the matrix. We append instead of overriding exisitng entry in this impl.

**Comment by nineinchnick** [comment]
> Question: Is it necessary to expand the matrix? Can we just passthrough the matrix as it is and let GHA handle the expansion?
> 
> e.g. in current impl we don't correctly handle include where it matches an already existing entry in the matrix. We append instead of overriding exisitng entry in this impl.
> 
> i.e. would it work if we just converted the `matrix.yaml` into a `matrix.json` and let it be?

Yes, that'll work, we just have to filter by modules in both the matrix and include sections. Thanks for pointing this out, I'm also not very confident that our matrix expansion would work exactly the same as GHA does it.

File: .github/bin/build-matrix-from-impacted.py [hunk 1]
```diff
@@ -38,7 +40,7 @@ def main():
         "--verbose",
         action="store_const",
         dest="loglevel",
-        const=logging.INFO,
+        const=logging.DEBUG,
```
Comment (hashhar) [inline .github/bin/build-matrix-from-impacted.py:43]: why? Do we get lot of noise from this?

Comment (nineinchnick) [reply .github/bin/build-matrix-from-impacted.py]: This is a leftover, reverted.

**Comment by nineinchnick** [comment]
> @nineinchnick Can you "test" by modifying the matrix to use in a commit:
> 
> The impl looks good to me % the case where includes overrides an existing module (a test would confirm).

I added tests in the script itself, PTAL. I'm not sure what you mean with overrides though.

**Review by hashhar — APPROVED** [review]
LGTM.

**Comment by hashhar** [comment]
> I'm not sure what you mean with overrides though.

I meant the case you added for `# include entry overwrites one in matrix`.

**Pull request merged**